packages/app/src/elements/audit.ts

```diff
@@ -0,0 +1,311 @@
+            return true;
+        }
+    }
+
+    return false;
+}
+
+export async function auditVaults(
+    vaults: Vault[],
+    {
+        updateOnlyItemWithId,
+        updateOnlyIfOutdated,
+    }: { updateOnlyItemWithId?: string; updateOnlyIfOutdated?: boolean } = {}
+) {
+    const reusedPasswords: ListItem[] = [];
+    const weakPasswords: ListItem[] = [];
+    const compromisedPasswords: ListItem[] = [];
+
+    const firstMatchPerPasswordHash = new Map<
+        string,
+        { item: VaultItem; vault: Vault; passwordField: { field: Field; fieldIndex: number } }
+    >();
+    const reusedPasswordItemIds: Set<string> = new Set();
+    const weakPasswordItemIds: Set<string> = new Set();
+    const compromisedPasswordItemIds: Set<string> = new Set();
+
+    const oneWeekAgo = sub(new Date(), { weeks: 1 });
+
+    // TODO: Remove this
+    console.log(`Running audit! ${JSON.stringify({ updateOnlyItemWithId, updateOnlyIfOutdated })}`);
+
+    for (const vault of vaults) {
+        for (const item of vault.items) {
+            if (updateOnlyItemWithId) {
+                if (item.id !== updateOnlyItemWithId) {
+                    // TODO: Remove this
+                    console.log("Skipped, not the matching item.");
+                    continue;
+                }
+            }
+
+            if (updateOnlyIfOutdated && item.lastAudited && item.lastAudited >= oneWeekAgo) {
+                // TODO: Remove this
+                console.log("Skipped, already up to date.");
+                continue;
+            }
+
+            const passwordFields = item.fields
+                .map((field, fieldIndex) => ({ field, fieldIndex }))
+                .filter((field) => field.field.type === FieldType.Password);
+
+            // If an item had password fields that failed audits and were since removed, we need to run the audit again to clear and update it
+            const itemHasFailedAudits = (item.auditResults || []).length > 0;
+
+            if (passwordFields.length === 0 && !itemHasFailedAudits) {
+                continue;
+            }
+
+            const auditResults: AuditResult[] = [];
+
+            for (const passwordField of passwordFields) {
+                const passwordHash = await sha1(passwordField.field.value);
+
+                // Perform reused audit
+                if (firstMatchPerPasswordHash.has(passwordHash)) {
+                    // Don't add the same item twice to the list, if there are more than one reused password fields in it
+                    if (!reusedPasswordItemIds.has(item.id)) {
+                        reusedPasswords.push({ item, vault });
+                        reusedPasswordItemIds.add(item.id);
+                    }
+
+                    auditResults.push({
+                        type: AuditResultType.ReusedPassword,
+                        fieldIndex: passwordField.fieldIndex,
+                    });
+
+                    // Also tag the first matching item as reused, once
```

Comment:
Why only tag the first matching item, and not all of them? If a password is reused accross more than 2 items, we'd want to tag all of them as reused, no?

Reply:
It would look strange to show the _same item_ multiple times on the same block (if it had more than one "failed" password field), that's why this check is in here. Do note it'll still show up in different blocks (weak and compromised), though.

Reply:
If an item has multiple reused password fields, both should be highlighted as such though, right? Besides, if you filter by `item.auditResults.some(res => res.type === AuditResultType.PasswordReused)` The item would still only appear once, even if it has multiple audit results of that type.

Reply:
All fields will be highlighted in the item view, not in the password audit. The password audit page shows the items in each bucket/box, so it doesn't matter if they have one or 10 fields failing the audit. Does that make sense, or are you thinking of doing something else?

We're not currently filtering like you mentioned, though, just adding items per bucket/box as the audit runs, so it's one less loop over everything.